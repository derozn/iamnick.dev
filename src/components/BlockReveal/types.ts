export type Props = {
  backgroundColor: string;
  delay?: number;
  show?: boolean;
  reverse?: boolean;
};

export type ContainerProps = Props & {
  entered: boolean;
};
