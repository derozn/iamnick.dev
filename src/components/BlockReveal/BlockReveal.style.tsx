import styled, { css, keyframes } from 'styled-components';
import { ContainerProps } from './types';

const scaleInOut = keyframes`
  0% {
    transform-origin: right center;
    transform: scale3d(0, 1, 1);
  }

  50% {
    transform-origin: right center;
    transform: scale3d(1, 1, 1);
  }
  51% {
    transform-origin: left center;
  }

  100% {
    transform-origin: left center;
    transform: scale3d(0, 1, 1);
  }
`;

const slideIn = keyframes`
  from {
    transform: translate3d(2%, 0, 0);
    opacity: 1;
  }
  to {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
`;

export const Container = styled.span<ContainerProps>`
  position: relative;
  display: inherit;

  &::after {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: ${({ backgroundColor }) => backgroundColor};
    transform: scale3d(0, 1, 1);
    ${({ entered, delay, reverse }) =>
      entered &&
      css`
        animation: ${scaleInOut} 1s ${delay ? delay + 0.5 : 0.5}s ease-in-out
          ${reverse && 'reverse'} forwards;
      `}
  }

  > span {
    display: inherit;
    opacity: 0;

    ${({ entered, delay, reverse }) =>
      entered &&
      css`
        animation: ${slideIn} 0.5s ${delay ? delay + 1 : 1}s ease-out ${reverse && 'reverse'}
          forwards;
      `}
  }
`;
