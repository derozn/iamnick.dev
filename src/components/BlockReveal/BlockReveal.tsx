import React, { useState, useEffect, FunctionComponent } from 'react';

import { Props } from './types';
import { Container } from './BlockReveal.style';

const BlockReveal: FunctionComponent<Props> = ({ children, show = true, ...rest }) => {
  const [entered, setEntered] = useState(show);

  useEffect(() => {
    if (!entered && show) setEntered(show);
  }, [entered, show]);

  return (
    <Container {...rest} entered={entered}>
      <span>{children}</span>
    </Container>
  );
};

export default BlockReveal;
