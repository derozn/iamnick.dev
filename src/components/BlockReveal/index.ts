export { default } from './BlockReveal';
export * from './types';
