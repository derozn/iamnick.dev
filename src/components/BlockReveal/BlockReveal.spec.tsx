import React from 'react';
import { render } from '@testing-library/react';

import BlockReveal from './BlockReveal';

describe('components/BlockReveal', () => {
  it('renders children', () => {
    const { getByText } = render(<BlockReveal backgroundColor="#fff">hello world!</BlockReveal>);

    expect(getByText('hello world!')).toBeTruthy();
  });

  it('renders with background css style set to passed in backgroundColor', () => {
    const { container } = render(<BlockReveal backgroundColor="#fff">hello world!</BlockReveal>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with additional delay factor on reveal animation', () => {
    const { container } = render(
      <BlockReveal backgroundColor="#fff" delay={0.5}>
        hello world!
      </BlockReveal>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
