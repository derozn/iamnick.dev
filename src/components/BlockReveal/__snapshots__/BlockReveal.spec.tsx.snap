// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BlockReveal renders with additional delay factor on reveal animation 1`] = `
.c0 {
  position: relative;
  display: inherit;
}

.c0::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: #fff;
  -webkit-transform: scale3d(0,1,1);
  -ms-transform: scale3d(0,1,1);
  transform: scale3d(0,1,1);
  -webkit-animation: hDiCPJ 1s 1s ease-in-out forwards;
  animation: hDiCPJ 1s 1s ease-in-out forwards;
}

.c0 > span {
  display: inherit;
  opacity: 0;
  -webkit-animation: jGGCbr 0.5s 1.5s ease-out  forwards;
  animation: jGGCbr 0.5s 1.5s ease-out  forwards;
}

<span
  class="c0"
>
  <span>
    hello world!
  </span>
</span>
`;

exports[`components/BlockReveal renders with background css style set to passed in backgroundColor 1`] = `
.c0 {
  position: relative;
  display: inherit;
}

.c0::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: #fff;
  -webkit-transform: scale3d(0,1,1);
  -ms-transform: scale3d(0,1,1);
  transform: scale3d(0,1,1);
  -webkit-animation: hDiCPJ 1s 0.5s ease-in-out forwards;
  animation: hDiCPJ 1s 0.5s ease-in-out forwards;
}

.c0 > span {
  display: inherit;
  opacity: 0;
  -webkit-animation: jGGCbr 0.5s 1s ease-out  forwards;
  animation: jGGCbr 0.5s 1s ease-out  forwards;
}

<span
  class="c0"
>
  <span>
    hello world!
  </span>
</span>
`;
